bin/varnishncsa/varnishncsa.c

```diff
@@ -163,8 +163,42 @@ static struct ctx {
 	int			recv_compl;
 } CTX;
 
+enum format_policy {
+        FMTPOL_FIRST = 0,
+        FMTPOL_REQ,
+        FMTPOL_RESP,
+};
+
 static void parse_format(const char *format);
 
+static unsigned
+process_format_policy(enum format_policy fp, unsigned first)
+{
+	unsigned policy;
+
+	if (fp == FMTPOL_FIRST)
+		return (first);
+
+#define BACKEND_FLAG 2<<8
+	policy = fp | ((*CTX.side == 'c') ? 0 : BACKEND_FLAG);
+
+	switch (policy) {
+		case FMTPOL_REQ:
+		case FMTPOL_RESP | BACKEND_FLAG:
+			if (CTX.recv_compl)
+				return (0);
+			return (first);
+		case FMTPOL_REQ | BACKEND_FLAG:
+			if (CTX.recv_compl)
+				return (0);
+			return (1);
+		case FMTPOL_RESP:
+			return (1);
+		default:
+			WRONG("Illegal enum value");
+	}
+}
```

Comment:
Putting it all together, this function should look like this instead:

```c
static unsigned
frag_needed(const struct fragment *frag, enum format_policy pol)
{
        unsigned is_first, want_first, want_frag;

        is_first = CTX.gen != frag->gen;

        switch (pol) {
        case FMTPOL_INTERNAL:
                want_first = 1;
                want_frag = 1;
                break;
        case FMTPOL_REQ:
                want_first = CTX.side == 'c';
                want_frag = !CTX.recv_compl;
                break;
        case FMTPOL_RESP:
                want_first = CTX.side == 'b';
                want_frag = (CTX.side == 'c') || !CTX.recv_compl;
                break;
        default:
                WRONG("Invalid format policy");
        }

        if (!want_frag)
                return (0);
        if (want_first && !is_first)
                return (0);
        return (1);
}
```

I have not tested it, but it should capture the semantics we established, and hopefully I'm not the only one thinking it's easier to read.